lib/internal/streams/from.js

```diff
@@ -34,21 +34,49 @@ function from(Readable, iterable, opts) {
     objectMode: true,
     ...opts
   });
+
   // Reading boolean to protect against _read
   // being called before last iteration completion.
   let reading = false;
+
+  // needToClose boolean if iterator needs to be explicitly closed
+  let needToClose = false;
+
   readable._read = function() {
     if (!reading) {
       reading = true;
       next();
     }
   };
+
+  readable._destroy = function(error, cb) {
+    close().then(
+      () => cb(error),
+      (e) => cb(error || e),
+    );
+  };
+
+  async function close() {
+    if (needToClose) {
+      needToClose = false;
+      if (typeof iterator.return === 'function') {
+        const { value } = await iterator.return();
+        await value;
+      }
+    }
+  }
+
   async function next() {
     try {
+      needToClose = false;
       const { value, done } = await iterator.next();
-      if (done) {
+      needToClose = !done;
+      const resolved = await value;
+      if (readable.destroyed) {
+        await close();
```

Comment:
Put this after `if(done)` since `if(done)` we don't need to explicitly close the iterator.

Reply:
Alternatively, `if(!done && readable.destroyed)`